# Linear Algebra with Numpy (Python)

In [233]:
from sympy.polys.subresultants_qq_zz import res
from sympy.stats.rv import probability
from tornado.autoreload import start
# install numpy
%pip install numpy

Note: you may need to restart the kernel to use updated packages.


In [234]:
import numpy as np

## Vectors

- Vector is a list of numbers or 1-D array in Numpy.
- Use vectors to represent: a feature column of a dataset; the gradient to descend the loss function

In [235]:
# create vectors in Numpy
x = np.array([0.5, 1.2, 3.5, 7.3])

In [236]:
print(x)

[0.5 1.2 3.5 7.3]


**Notice:** `x` is a object in Numpy.

In [237]:
# check the type of the vector
print(type(x))

<class 'numpy.ndarray'>


For any `Numpy` array, we have some properties: `x.shape` (shape of the array) or `x.ndim` (dimensions) and `x.dtype` (data type)

In [238]:
print(x.shape)

(4,)


In [239]:
print(x.ndim)

1


In [240]:
print(x.dtype)

float64


In [241]:
x = np.array([0.5, 1.2, 3.5, 7.3])
y = np.array([0.2, 1.4, 2.2, 2.1])

In [242]:
len(x) == len(y)

True

In [243]:
# linear combination of vectors
z = 2*x + 3*y
print(z)

[ 1.6  6.6 13.6 20.9]


**Elementwise product:**

If $x = (x_1,x_2,\ldots,x_n)$ and $y = (y_1,y_2,\ldots,y_n)$ then
$$
x*y = (x_1y_1, x_2y_2, \ldots, x_ny_n).
$$

In [244]:
# what is x * y?
print(x*y)

[ 0.1   1.68  7.7  15.33]


**Dot Product:**
If $x = (x_1,x_2,\ldots,x_n)$ and $y = (y_1,y_2,\ldots,y_n)$ then
$$
x \dot y = x_1y_1 + x_2y_2+\cdots+x_ny_n.
$$

In [245]:
# compute the dot product
np.dot(x,y) == np.sum(x*y)

np.True_

In [246]:
x @ y

np.float64(24.810000000000002)

## Norm (len) of vectors
A norm of a vector is defined as
$$
\|x\|_{p} = (|x_1|^p+|x_2|^p+\cdots+|x_n|^p)^{1/p}
$$
In Numpy, we only need for $p=1$, $p=2$ or $p=\infty$

In [247]:
# compute the norm of vector
v = np.array([3.0, -4.0, 0])

In [248]:
l1_norm = np.linalg.norm(v, 1)
print(l1_norm)

7.0


In [249]:
linf_norm = np.linalg.norm(v, np.inf)
print(linf_norm)

4.0


In [250]:
# 12 norm
l2_norm = np.linalg.norm(v, 2)
print(l2_norm)

5.0


In [251]:
l2_norm = np.linalg.norm(v)
print(l2_norm)

5.0


**Example:** Write a function to compute the cosine similarity of two vectors.
If $x = (x_1,\ldots,x_n)$ and $y = (y_1,\ldots,y_n)$ then
$$
\cos(\theta) = \frac{x\cdot y}{\|x\|\|y\|}.
$$

In [252]:
def cos_sim(x, y):
    return (x @ y) / (np.linalg.norm(x) * np.linalg.norm(y))

In [253]:
# make up vectors
cat = np.array([0.9, 0.1, 0.0])
dog = np.array([0.85, 0.2, 0.0])
car = np.array([0.0, 0.1, 0.9])

In [254]:
# check the similarity between cat and dog
cos_sim(cat, dog)

np.float64(0.9927566538132715)

In [255]:
cos_sim(cat, car)

np.float64(0.012195121951219513)

## Matrices

Matrix is just a 2-D array in Numpy (or a list of lists). Use matrices for
- dataset (rows and columns)
- weight of layers in a DL model
- a batch of vectors to feed into a model.

In [256]:
# create a matrix with Numpy
A = np.array([
    [1.0, 2.0, 3.0],
    [4.0, 5.0, 6.0],
])

In [257]:
print(A)

[[1. 2. 3.]
 [4. 5. 6.]]


In [258]:
print(A.shape)

(2, 3)


In [259]:
print(A.ndim)

2


In [260]:
print(A.dtype)

float64


**Operations on matrices:**

- Transpose of a matrix
- Indexing or slicing
- Matrix multiplication

In [261]:
A = np.array([
    [1.0, 2.0, 3.0],
    [4.0, 5.0, 6.0],
])

In [262]:
# create a matrix with Numpy
B = np.array([
    [1.4, 1.6, 3.7],
    [4.5, -5.1, 2.2],
])

In [263]:
# compute the transpose
A.T

array([[1., 4.],
       [2., 5.],
       [3., 6.]])

In [264]:
np.transpose(A)

array([[1., 4.],
       [2., 5.],
       [3., 6.]])

In [265]:
print(A.shape)
print(B.shape)

(2, 3)
(2, 3)


In [266]:
# Matrix multiply
A @ B.T

array([[15.7,  0.9],
       [35.8,  5.7]])

**Indexing or slicing:**

$A[i]$ gives the row i-th, $A[:, j]$ gives the column j-th; $A[i,j]$ gives the element at the position (i, j).

In [267]:
A

array([[1., 2., 3.],
       [4., 5., 6.]])

In [268]:
A[0]

array([1., 2., 3.])

In [269]:
A[1]

array([4., 5., 6.])

In [270]:
# 1st col
A[:,0]

array([1., 4.])

In [271]:
# if we want to get value 5, then
A[1,1]

np.float64(5.0)

## Rank and determinate of matrices

In [272]:
A = np.array([
    [1.1, 2.1, 2.5],
    [2.2, 3.1, -1.1],
    [0.5, 0.1, -0.3]
])

In [273]:
rank = np.linalg.matrix_rank(A)

In [274]:
print(rank)

3


In [275]:
# compute the determinant
det = np.linalg.det(A)
print(det)

-3.9959999999999996


In [276]:
A_inv = np.linalg.inv(A)

In [277]:
print(A_inv)

[[ 0.20520521 -0.22022022  2.51751752]
 [-0.02752753  0.3953954  -1.67917918]
 [ 0.33283283 -0.23523524  0.3028028 ]]


In [278]:
A @ A_inv

array([[ 1.00000000e+00, -5.55111512e-17,  8.32667268e-17],
       [ 2.03985422e-17,  1.00000000e+00, -1.32365479e-16],
       [ 1.47974170e-17,  1.40417096e-17,  1.00000000e+00]])

In [279]:
# generate an identity matrix
I_3 = np.eye(3)

In [280]:
I_3

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [281]:
# Check if A@A_inv == I_3 (identity)
np.allclose(A@A_inv, I_3)

True

## Special vectors and matrices

In [282]:
# zero vectors
x = np.zeros(5)
print(x)

[0. 0. 0. 0. 0.]


In [283]:
# vector with ones
y = np.ones(5)
print(y)

[1. 1. 1. 1. 1.]


In [284]:
# create a vector of type float
a = np.array([1, 2, 3, 4, 5], dtype=float)
print(a)

[1. 2. 3. 4. 5.]


In [285]:
# create a zero vector that has the same len and same dtypes as a
b = np.array([0, 0, 0, 0, 0], dtype=float)
print(b)

[0. 0. 0. 0. 0.]


In [286]:
# list comprehension
b = np.array([0 for i in range(len(a))], dtype=float)
print(b)

[0. 0. 0. 0. 0.]


In [287]:
# use zero function
b = np.zeros(len(a), dtype=float)
print(b)

[0. 0. 0. 0. 0.]


In [288]:
# Pythonic way to do this
b = np.zeros_like(a)
print(b)

[0. 0. 0. 0. 0.]


In [289]:
# create zero matrix
A = np.zeros((3,4)) # (3,4) is the shape
print(A)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


## Randomization in Numpy

**Generate random vectors:**

In [290]:
# generate random integer vector
x = np.random.randint(low=1, high=5, size=5)
print(x)

[3 1 2 2 3]


In [291]:
# gen random float vector
y = np.random.normal(loc=0, scale=1.4, size=6)
print(y)

[-0.7252448  -1.99305787 -4.32534465 -1.82763517 -2.37332689 -2.38019486]


In [292]:
# generate random float vector (uniform distribution)
z = np.random.uniform(low=0, high=3, size=5)
print(z)

[1.91310845 1.25206059 0.33840096 1.37200873 0.05011634]


**Generate random matrices:**

In [293]:
# gen int matrix
A = np.random.randint(low=1, high=5, size=(4,5))
print(A)

[[3 2 1 4 2]
 [3 3 3 4 4]
 [4 3 3 3 1]
 [1 3 2 2 3]]


In [294]:
# gen random float matrix (normal distributin)
B = np.random.normal(loc=1.0, scale=0.5, size=(4,4))
print(B)

[[0.81298244 1.01966554 0.34675143 0.45042226]
 [1.29806224 1.66372069 1.63884928 1.12795105]
 [0.49324509 0.9413897  0.18393702 1.23474877]
 [1.32037825 1.50845855 1.03366438 0.90313843]]


## Vector-valued function
Suppose we have a function $y=f(x)$ and a vector $v = (v_1,\ldots, v_n)$. We want to define $f(v)$.

$$
f(v) = (f(v_1),\ldots, f(v_1)).
$$

**Example:**
- Generate a random vector, v, (float) of size n = 200
- Use the for loop to determine the sin(v).

In [295]:
# compute sin(v)
import math

In [296]:
n = 5000
# gen random vector (either normal or uniform distribution)
v = np.random.normal(loc=0, scale=1.5, size=n)
print(v.shape)

(5000,)


In [297]:
# use for loop
# write a function to compute sin_v
def sin(v):
    result = []
    n = len(v)
    for i in range(n):
        result.append(math.sin(v[i]))
    sin_v = np.array(result)
    return sin_v

In [298]:
print(sin(v).shape)

(5000,)


**Measure the running time**

In [299]:
import time

In [300]:
# test the computing time for n from 5000 to 10000
for n in range(5_000, 10_000, 100):
    v = np.random.normal(loc=0, scale=1.5, size=n)
    star_time = time.time()
    sin_v = sin(v)
    end_time = time.time()
    print(f"The running time for n={n} is {end_time - star_time}")

The running time for n=5000 is 0.0008649826049804688
The running time for n=5100 is 0.0008630752563476562
The running time for n=5200 is 0.0009000301361083984
The running time for n=5300 is 0.0008409023284912109
The running time for n=5400 is 0.0008769035339355469
The running time for n=5500 is 0.0009038448333740234
The running time for n=5600 is 0.0008800029754638672
The running time for n=5700 is 0.0008840560913085938
The running time for n=5800 is 0.0008969306945800781
The running time for n=5900 is 0.0009200572967529297
The running time for n=6000 is 0.0009300708770751953
The running time for n=6100 is 0.0009992122650146484
The running time for n=6200 is 0.0010178089141845703
The running time for n=6300 is 0.0010020732879638672
The running time for n=6400 is 0.0009970664978027344
The running time for n=6500 is 0.001004934310913086
The running time for n=6600 is 0.0010209083557128906
The running time for n=6700 is 0.0010390281677246094
The running time for n=6800 is 0.00106382369995

In [301]:
n = 10_000_000
v = np.random.normal(loc=0, scale=1.5, size=n)
star_time = time.time()
sin_v = sin(v)
end_time = time.time()
print(f"Running Time for: {end_time - star_time}")

Running Time for: 1.6221072673797607


**Vectorization:** In Numpy, we can compute $f(v)$ at once using the function in Numpy. They are vectorized already.

In [302]:
import numpy as np
import time

In [303]:
n = 10_000_000

In [304]:
v = np.random.normal(loc=0, scale=1.5, size=n)
star_time = time.time()
# use numpy directly
sin_v = np.sin(v)
end_time = time.time()
print(f"Running Time for: {end_time - star_time}")

Running Time for: 0.09165477752685547


In [305]:
# benchmark these approaches
for n in range(10_000_000, 12_000_001, 1_000_000):
    v = np.random.normal(loc=0, scale=1.5, size=n)
    star_time = time.time()
    # compute the running time using normal loop
    sin_v = sin(v)
    end_time = time.time()
    print(f"Running time for pure loop: {end_time - star_time}")
    # reset the time
    star_time = time.time()
    # compute using vectorized numpy
    sin_v = np.sin(v)
    end_time = time.time()
    print(f"Running time for vectorization: {end_time - star_time}")

Running time for pure loop: 1.6195440292358398
Running time for vectorization: 0.09098386764526367
Running time for pure loop: 1.7901382446289062
Running time for vectorization: 0.12068915367126465
Running time for pure loop: 1.9468989372253418
Running time for vectorization: 0.11179494857788086


Two functions in your homework: `Relu` and `softmax`.

$Relu(x) = \max(0,x)$. If $x > 0$ then $Relu(x) = x$; otherwise $Relu(x) = 0.$

In [355]:
def relu(x):
    x = np.array(x)
    zero = np.zeros_like(x)
    return np.max(np.array([zero, x]), axis=0)

In [307]:
def relu2(x):
    return max(0,x)

In [356]:
# test them
x = -0.5
print(relu(x))
print(relu2(x))

0.0
0


In [357]:
x = [1, -2]
print(relu(x))

[1 0]


In [310]:
x = np.random.normal(loc=0, scale=2, size=10)
print(x)

[-0.87654066 -0.85171402  0.56011761 -2.25366162  4.22678783 -1.50619162
  1.99125591 -0.34267692  0.44002424 -0.59886485]


In [358]:
relu(x)

array([1, 0])

Define `softmax` function: Turn a vector into probabilies. If the value inside the vector id larger, it must have a bigger probability.

Math formula:
$$
softmax(v) = \frac{(e^{v_1},\ldots,e^{v_n})}{e^{v_1}+\cdots+e^{v_n}}
$$

In [312]:
# define vectorized function softmax
def softmax(v):
    # if c is not a numpy array
    v = np.array(v)
    e_v = np.exp(v)
    return e_v / np.sum(e_v)

In [313]:
# test the function
v = np.random.normal(loc=0, scale=2.0, size=10)
print(v)

[-3.21829745  2.06815598 -1.41550842 -2.90425962 -2.72217256  0.28693946
 -2.92869929  2.53481409 -1.88834354 -5.03205334]


In [314]:
softmax(v).round(3)

array([0.002, 0.352, 0.011, 0.002, 0.003, 0.059, 0.002, 0.561, 0.007,
       0.   ])

**Classification problems:**

Suppose we have 3 types of fruits: apples, banana, and orange. Each fruit has a presentation as a vector of 4 dimensions. Below is the data set for `[apple, apple, banana, banana, orange, orange]`

In [317]:
fruit_labels = ["apple", "apple", "banana", "banana", "orange", "orange"]

In [329]:
fruit_vectors = np.array([
    [0.78, 0.5, 0.88, 0.65],
    [0.79, 0.39, 0.9, 0.69],
    [0.08, 0.94, 0.82, 0.78],
    [0.07, 0.96, 0.79, 0.8],
    [0.15, 0.47, 0.48, 0.91],
    [0.1, 0.41, 0.4, 0.98]
])
fruit_vectors

array([[0.78, 0.5 , 0.88, 0.65],
       [0.79, 0.39, 0.9 , 0.69],
       [0.08, 0.94, 0.82, 0.78],
       [0.07, 0.96, 0.79, 0.8 ],
       [0.15, 0.47, 0.48, 0.91],
       [0.1 , 0.41, 0.4 , 0.98]])

Suppose we have a new fruit (one of the 3 given) with vector

In [335]:
new_fruit_vector = np.array([0.05, 0.93, 0.79, 0.82])

Determine the name of the new fruit.

In [319]:
# define cosine similarity score
def cos_score(a,b):
    return np.dot(a,b) / (np.linalg.vector_norm(a) * np.linalg.vector_norm(b))

In [320]:
a = np.random.normal(loc=0, scale=1.5, size=5)
print(a)

[-0.31771558  1.2341355   0.73679694 -0.54188884 -2.18162219]


In [321]:
b = np.random.normal(loc=0, scale=1.5, size=5)
print(b)

[ 3.37214538 -0.82472641 -2.19456621  0.50400853 -0.68614453]


In [322]:
# test cos_score function
score = cos_score(a,b)
print(score)

-0.2202586316142076


To identify the new fruit, we compare the cosine scores of the new vectors paring with the known fruit.

In [323]:
first_apple = fruit_vectors[0]
print(first_apple)

[0.78, 0.5, 0.88, 0.65]


In [324]:
score1 = cos_score(first_apple,new_fruit_vector)
print(score1)

0.8213756609884086


In [337]:
# better to use loop
for i, v in enumerate(fruit_vectors):
    # i -> 0, 1, 2, 3, 4, 5
    # v -> fruit vectors in order
    score_i = cos_score(new_fruit_vector, v)
    print(f"The score of the new fruit vs {fruit_vectors[i]} is {score_i}")

The score of the new fruit vs [0.78 0.5  0.88 0.65] is 0.8213756609884086
The score of the new fruit vs [0.79 0.39 0.9  0.69] is 0.7947969315916223
The score of the new fruit vs [0.08 0.94 0.82 0.78] is 0.999193828247798
The score of the new fruit vs [0.07 0.96 0.79 0.8 ] is 0.9996283906419465
The score of the new fruit vs [0.15 0.47 0.48 0.91] is 0.9353889721593739
The score of the new fruit vs [0.1  0.41 0.4  0.98] is 0.8983727348814865


Based on the output scores, the new fruit is banana

In [340]:
# compute all scores at once
scores = []
for i in range(6):
    score = cos_score(new_fruit_vector, fruit_vectors[i])
    scores.append(score)
# convert list score to Numpy
scores = np.array(scores)

In [341]:
scores

array([0.82137566, 0.79479693, 0.99919383, 0.99962839, 0.93538897,
       0.89837273])

Find the index where the score is max

In [342]:
max_idx = np.argmax(scores)
print(max_idx)

3


In [344]:
label = fruit_labels[max_idx]
print(label)

banana


## Small neural network!

$$
XW_1 + b_1 \rightarrow ReLU \rightarrow HW_2 + b_2 \rightarrow softmax \rightarrow class
$$
where $W_1, W_2$ are to matrices (weights) and two vectors $b_1, b_2$ (biases); the activation function is $ReLU$ and $softmax$ is the probability function.

In [345]:
W_1 = np.array(
    [
        [-4.91, -0.09, 0.31],
        [6.06, -0.28, 0.17],
        [2.21, 0.45, -0.36],
        [-0.90, -0.33, 0.01],
    ]
)
print(W_1)

[[-4.91 -0.09  0.31]
 [ 6.06 -0.28  0.17]
 [ 2.21  0.45 -0.36]
 [-0.9  -0.33  0.01]]


In [346]:
W_2 = np.array(
    [
        [-7.04, 3.99, 1.16],
        [-0.36, -0.27, -0.16],
        [0.21, 0.52, -0.07],
    ]
)
print(W_2)

[[-7.04  3.99  1.16]
 [-0.36 -0.27 -0.16]
 [ 0.21  0.52 -0.07]]


In [349]:
b_1 = np.array([-0.55, -0.03, -0.01])
b_2 = np.array([7.13, -8.65, 1.52])

In [350]:
X = new_fruit_vector

In [352]:
# compute XW_1 + b_1 -> y_1
y_1 = X@W_1 + b_1

In [353]:
print(y_1)

[ 5.8482 -0.21   -0.1126]


In [360]:
# compute H
H = relu(y_1)
print(H)

[5.8482 0.     0.    ]


In [361]:
# compute HW_2 + b_2 -> y_2
y_2 = H@W_2 + b_2

In [362]:
print(y_2)

[-34.041328  14.684318   8.303912]


In [366]:
# apply the softmax to get the probability
prob = softmax(y_2)
print(prob.round(3))

[0.    0.998 0.002]


Notice: The probability vector has 3 values which are the probability for the class: apple, banana and orange in this order.

probabiluity( new fruit os apple) = 0

probabiluity( new fruit os banana) = 0.998

probabiluity( new fruit os orange) = 0.002

In [367]:
classes = ["apple", "banana", "orange"]

In [368]:
max_idx = np.argmax(prob)
print(max_idx)

1


In [369]:
fruit = classes[max_idx]
print(fruit)

banana


In [370]:
fruit_vectors

array([[0.78, 0.5 , 0.88, 0.65],
       [0.79, 0.39, 0.9 , 0.69],
       [0.08, 0.94, 0.82, 0.78],
       [0.07, 0.96, 0.79, 0.8 ],
       [0.15, 0.47, 0.48, 0.91],
       [0.1 , 0.41, 0.4 , 0.98]])

In [372]:
X = fruit_vectors[5:]
print(X)

[[0.1  0.41 0.4  0.98]]


In [373]:
y_1 = X@W_1 + b_1
H = relu(y_1)
y_2 = H@W_2 + b_2
prob = softmax(y_2)
max_idx = np.argmax(prob)
print(classes[max_idx])

orange
